File: frontend/src/types/monitoring.ts
export interface MonitoringKPI {
  total_calls: number
  avg_latency_ms: number
  p95_latency_ms: number
  total_tokens: number
  estimated_cost: number
  error_count: number
  error_rate: number
  prev_total_calls: number
}

export interface TokenTrendPoint {
  date: string
  prompt_tokens: number
  completion_tokens: number
}

export interface LatencyBucket {
  range: string
  count: number
}

export interface ScoreAgreement {
  agree_pct: number
  minor_diff_pct: number
  major_divergence_pct: number
  total: number
}

export interface CostBreakdownPoint {
  date: string
  model: string
  cost: number
}

export interface RecentLLMAnalysis {
  time: string
  case_number: number | null
  sender: string | null
  llm_score: number | null
  final_score: number | null
  tokens: number | null
  latency_ms: number | null
  status: string
  model?: string | null
  threat_category?: string | null
  explanation_summary?: string | null
}

export interface MonitoringData {
  kpi: MonitoringKPI
  token_trend: TokenTrendPoint[]
  latency_distribution: LatencyBucket[]
  score_agreement: ScoreAgreement
  cost_breakdown: CostBreakdownPoint[]
  recent_analyses: RecentLLMAnalysis[]
}

// ML Monitoring Types
export interface MLMonitoringKPI {
  total_calls: number
  avg_latency_ms: number
  p95_latency_ms: number
  avg_confidence: number
  error_count: number
  error_rate: number
  prev_total_calls: number
}

export interface ScoreDistributionBucket {
  range: string
  count: number
}

export interface ConfidenceAccuracyPoint {
  confidence: number
  accuracy: number
}

export interface LatencyTrendPoint {
  date: string
  avg_latency_ms: number
}

export interface RecentMLAnalysis {
  time: string
  case_number: number | null
  sender: string | null
  ml_score: number | null
  final_score: number | null
  confidence: number | null
  latency_ms: number | null
  status: string
}

export interface MLMonitoringData {
  kpi: MLMonitoringKPI
  score_distribution: ScoreDistributionBucket[]
  confidence_accuracy: ConfidenceAccuracyPoint[]
  latency_trend: LatencyTrendPoint[]
  score_agreement: ScoreAgreement
  recent_analyses: RecentMLAnalysis[]
}

// Heuristics Monitoring Types
export interface HeuristicsMonitoringKPI {
  total_runs: number
  avg_latency_ms: number
  p95_latency_ms: number
  high_score_rate: number
  zero_score_rate: number
  prev_total_runs: number
}

export interface TopRulePoint {
  rule_name: string
  count: number
}

export interface RecentHeuristicsAnalysis {
  time: string
  case_number: number | null
  sender: string | null
  heuristic_score: number | null
  final_score: number | null
  triggered_rules: string[]
  latency_ms: number | null
}

export interface HeuristicsMonitoringData {
  kpi: HeuristicsMonitoringKPI
  top_triggered_rules: TopRulePoint[]
  score_distribution: ScoreDistributionBucket[]
  score_agreement: ScoreAgreement
  recent_analyses: RecentHeuristicsAnalysis[]
}

File: backend/app/schemas/monitoring.py
"""Monitoring page response schemas."""

from datetime import datetime

from pydantic import BaseModel


class MonitoringKPI(BaseModel):
    total_calls: int
    avg_latency_ms: float
    p95_latency_ms: float
    total_tokens: int
    estimated_cost: float
    error_count: int
    error_rate: float
    prev_total_calls: int


class TokenTrendPoint(BaseModel):
    date: str
    prompt_tokens: int
    completion_tokens: int


class LatencyBucket(BaseModel):
    range: str
    count: int


class ScoreAgreement(BaseModel):
    agree_pct: float
    minor_diff_pct: float
    major_divergence_pct: float
    total: int


class CostBreakdownPoint(BaseModel):
    date: str
    model: str
    cost: float


class RecentLLMAnalysis(BaseModel):
    time: datetime
    case_number: int | None = None
    sender: str | None = None
    llm_score: float | None = None
    final_score: float | None = None
    tokens: int | None = None
    latency_ms: int | None = None
    status: str
    model: str | None = None
    threat_category: str | None = None
    explanation_summary: str | None = None

    model_config = {"from_attributes": True}


class MonitoringResponse(BaseModel):
    kpi: MonitoringKPI
    token_trend: list[TokenTrendPoint] = []
    latency_distribution: list[LatencyBucket] = []
    score_agreement: ScoreAgreement
    cost_breakdown: list[CostBreakdownPoint] = []
    recent_analyses: list[RecentLLMAnalysis] = []


# ML Monitoring Schemas
class MLMonitoringKPI(BaseModel):
    total_calls: int
    avg_latency_ms: float
    p95_latency_ms: float
    avg_confidence: float
    error_count: int
    error_rate: float
    prev_total_calls: int


class ScoreDistributionBucket(BaseModel):
    range: str
    count: int


class ConfidenceAccuracyPoint(BaseModel):
    confidence: float
    accuracy: float


class LatencyTrendPoint(BaseModel):
    date: str
    avg_latency_ms: float


class RecentMLAnalysis(BaseModel):
    time: datetime
    case_number: int | None = None
    sender: str | None = None
    ml_score: float | None = None
    final_score: float | None = None
    confidence: float | None = None
    latency_ms: int | None = None
    status: str

    model_config = {"from_attributes": True}


class MLMonitoringResponse(BaseModel):
    kpi: MLMonitoringKPI
    score_distribution: list[ScoreDistributionBucket] = []
    confidence_accuracy: list[ConfidenceAccuracyPoint] = []
    latency_trend: list[LatencyTrendPoint] = []
    score_agreement: ScoreAgreement
    recent_analyses: list[RecentMLAnalysis] = []


# Heuristics Monitoring Schemas
class HeuristicsMonitoringKPI(BaseModel):
    total_runs: int
    avg_latency_ms: float
    p95_latency_ms: float
    high_score_rate: float
    zero_score_rate: float
    prev_total_runs: int


class TopRulePoint(BaseModel):
    rule_name: str
    count: int


class RecentHeuristicsAnalysis(BaseModel):
    time: datetime
    case_number: int | None = None
    sender: str | None = None
    heuristic_score: float | None = None
    final_score: float | None = None
    triggered_rules: list[str] = []
    latency_ms: int | None = None

    model_config = {"from_attributes": True}


class HeuristicsMonitoringResponse(BaseModel):
    kpi: HeuristicsMonitoringKPI
    top_triggered_rules: list[TopRulePoint] = []
    score_distribution: list[ScoreDistributionBucket] = []
    score_agreement: ScoreAgreement
    recent_analyses: list[RecentHeuristicsAnalysis] = []


# Score Analysis Schemas
class ScoreMetrics(BaseModel):
    agreement_rate: float
    avg_std_dev: float
    correlation_heur_ml: float
    correlation_heur_llm: float
    correlation_ml_llm: float
    total_cases_analyzed: int


class CaseScoreBreakdown(BaseModel):
    case_number: int
    email_sender: str
    email_received_at: datetime
    heuristic_score: float | None
    ml_score: float | None
    llm_score: float | None
    final_score: float
    std_dev: float
    agreement_level: str

    model_config = {"from_attributes": True}


class ScoreAnalysisResponse(BaseModel):
    metrics: ScoreMetrics | None
    cases: list[CaseScoreBreakdown]
